src/actions/filesDropped/metadata.js

```diff
@@ -1,124 +1,248 @@
-  let csvData, errors, csvMeta;
+  const fileName = file.name;
+
   try {
-    ({data: csvData, errors, meta: csvMeta} = await parseCsv(file));
+    /* Parse & interrogate the CSV file */
+    const {errors, data, meta} = await parseCsvTsv(file);
     if (errors.length) {
       console.error(errors);
       throw new Error(errors.map((e) => e.message).join(", "));
     }
+    const {coloringInfo, strainKey, latLongKeys, ignoredFields} = processHeader(meta.fields);
+    const rows = {};
+    data.forEach((d) => {rows[d[strainKey]]=d;});
+
+    /* For each coloring, extract values defined in each row etc */
+    const newNodeAttrs = {};
+    const newColorings = processColorings(newNodeAttrs, coloringInfo, rows, fileName); // modifies `newNodeAttrs`
+    const newGeoResolution = latLongKeys ? processLatLongs(newNodeAttrs, latLongKeys, rows, fileName) : undefined;
+    /* Fix errors in data & dispatch warnings here, as we cannot dispatch in the reducers */
+    const ok = checkDataForErrors(dispatch, getState, newNodeAttrs, newColorings, ignoredFields, fileName);
+    if (!ok) return undefined;
+
+    dispatch({type: ADD_EXTRA_METADATA, newColorings, newGeoResolution, newNodeAttrs});
+    return dispatch(successNotification({
+      message: `Adding metadata from ${fileName}`,
+      details: `${Object.keys(newColorings).length} new coloring${Object.keys(newColorings).length > 1 ? "s" : ""} for ${Object.keys(newNodeAttrs).length} node${Object.keys(newNodeAttrs).length > 1 ? "s" : ""}`
+    }));
+
   } catch (err) {
     return dispatch(errorNotification({
-      message: `Parsing of ${file.name} failed`,
+      message: `Parsing of ${fileName} failed`,
       details: err.message
     }));
   }
+};
 
-  const strainKey = csvMeta.fields[0];
-  const {controls, tree} = getState();
-  const newColorByNames = [];
-  const colorBysIgnored = [];
-  /* currently we cannot process metadata fields which are defined as properties on a node, rather than trait properties
-  These could be included on a case-by-case basis */
-  const fieldsToIgnore = new Set(["name", "div", "num_date", "vaccine", "labels", "hidden", "mutations", "url", "authors", "accession", "traits", "children"]);
-  csvMeta.fields.slice(1).forEach((colorBy) => {
-    if (controls.coloringsPresentOnTree.has(colorBy) || fieldsToIgnore.has(colorBy)) {
-      colorBysIgnored.push(colorBy);
-    } else {
-      newColorByNames.push(colorBy);
+export default handleMetadata;
+
+
+/* ---------------------- helper functions to parse data ---------------------------- */
+
+
+function processHeader(fields) {
+  const strainKey = fields[0];
+
+  /* There are a number of "special case" columns we currently ignore */
+  const fieldsToIgnore = new Set(["name", "div", "vaccine", "labels", "hidden", "mutations", "url", "authors", "accession", "traits", "children"]);
+  fieldsToIgnore.add("num_date").add("year").add("month").add("date"); /* TODO - implement date parsing */
```

Comment:
An obvious next step is to allow dates to be specified via additional metadata. Interpreting this as the `num_date` attr is slightly complex as the coloring & tree-metric are tied together in the code. A different approach would be to extend our color schemes to allow "temporal" types and then add temporal info in the CSV/TSV as simply an additional color-by.

Reply:
So, this would mean sequences in a time-tree with inferred dates were allowed to have 'real' dates supplied? Would it change where they're plotted on X-axis?

Reply:
I'm leaning towards allowing metadata traits (in drag-n-drop data or in the main JSON) which are encoded as (e.g.) `YYYY-MM-DD` to be inferred as temporal types by auspice, and a colour scale generated according. A good example of this would be nCoV's "submitted date" field. These would be "just another color-by" and thus different to the `num_date` field which is used by auspice for the tree-time view. I don't see any easy way to use metadata to update / influence node (temporal) positioning, but am open to suggestions. (It's easy to see how such metadata could define the x-axis position of tips, but what do we do with internal nodes / tips not in the metadata?)

Reply:
Sounds good! Yes, I agree, that's why I was curious - was wondering how you'd adjust the other parts of the tree to accommodate moving tips! Unfortunately I have no answers to offer, just curiosity :)